### PR TITLE
When a task fails and doesn't throw an exception, report it correctly…

### PR DESCRIPTION
**Pull request by churromorales**
Noticed when running in MM-less mode, that when a task failed and then the runTask() method would swallow the exception.  Thus it was reported as a success when it had in fact failed.  This fixes that issue.

### CODE REVIEW

**Comment by kfaraz** [comment]
Thanks for the fix, @churromorales !

Taking a look at the original changes, I wonder if we need the `UpdateStatusAction`. It is invoked only when the task has finished, either successfully or not, and it leads to the triggering of the registered `TaskRunnerListener`s. 

Rather than creating a new task action, I think it would be cleaner to just update it inside `KubernetesTaskRunner`, which already seems to be happening. This would also help us get rid of the `TaskRunner.updateStatus()` which really seems to be in contradiction with the `TaskRunner` API. We shouldn't have to tell the `TaskRunner` of the status of the task since it returns the status of the task itself in `ListenableFuture<TaskStatus> run(Task task)`.

Does triggering the action from `AbstractTask` satisfy a specific corner case which is not covered by the `KubernetesTaskRunner`?

**Comment by churromorales** [comment]
@kfaraz great question.  Originally that was the design, but when testing it in reality I encountered quite a few problems doing it this way.  I had a comment in the original PR describing why I chose to go with handling it in the AbstractTask: 

https://github.com/apache/druid/pull/13156#issuecomment-1278289541

Let me know if you still have questions. 
Thank you

**Comment by churromorales** [comment]
@kfaraz I was wondering if you were okay with the response.  

TLDR: 

When a k8s task completes, we monitor the job's exit code to see if we were successful or not.  The KubernetesTaskRunner does this as a fallback in case the peon's message didn't arrive back to the overlord properly. 

But the real issue is that the time between when the java (peon) process terminates, and the pod gracefully shuts down (which usually has some terminiation grace period for a preStop hook to execute) to when the job is marked as success or failure is not trivial in many situations.  And if these tasks are holding locks, then we are holding locks for longer than we should.  Let me know if you have any further questions on this approach.

**Comment by churromorales** [comment]
> 
the task runner before was forking the task but on the same host.  Now you don't have that luxury, you had all the info on your local filesystem and control of the process, now you don't have that luxury anymore.

**Review by abhishekagarwal87 — APPROVED** [review]

**Review by a2l007 — APPROVED** [review]
LGTM

**Comment by churromorales** [comment]
🙏

**Comment by kfaraz** [comment]
Thanks for the clarification, @churromorales ! 
Sorry, I was out last week and couldn't respond sooner.
